File: Sorting/Python/selection.py
from typing import Any, List
import random

def timetaken(fn):
    def wrapper(*args, **kwargs):
        from time import time
        start = time()
        result = fn(*args, **kwargs)
        end = time()
        print(f"Time Taken: {end - start} sec")
        return result
    return wrapper


class Selection:
    @staticmethod
    @timetaken
    def sort(items: Any):
        n = len(items)
        for i in range(n-1):
            min_index = i
            for j in range(i+1, n):
                if Selection.less(items[j], items[min_index]):
                    min_index = j
            Selection.exchange(items, i, min_index)

    @staticmethod
    def less(a: Any, b: Any) -> bool:
        return a < b

    @staticmethod
    def exchange(items: List, i, j):
        items[i], items[j] = items[j], items[i]


if __name__ == '__main__':
    items: List[int] = [random.randint(0, 50000) for x in range(20000)]
    Selection.sort(items)
    Selection.sort(items)

File: Sorting/Java/Selection.java
public class Selection {
    public static void sort(Comparable[] items){
        int N = items.length;
        for (int i =0; i < N-1; i++){
            int min = i+1;
            for (int j = i+1; j < N; j++){
                if(less(items[j], items[i])){
                    min = j;
                }
                exchange(items, i, min);
            }
        }
    }

    public static boolean less(Comparable a, Comparable b){
        return a.compareTo(b) < 0;
    }

    public static void exchange(Comparable[] items, int i, int j){
        Comparable temp = items[i];
        items[i] = items[j];
        items[j] = temp;
    }

    public static void main(String[] args){
        int[] items = {5, 4, 3, 2, 1};
        Selection.sort(items);
        System.out.println(items);
    }
}